server/src/infra/infra.utils.ts

```diff
@@ -122,6 +125,77 @@ export function ChunkedSet(options?: { paramIndex?: number }): MethodDecorator {
   return Chunked({ ...options, mergeFn: setUnion });
 }
 
+// https://stackoverflow.com/a/74898678
+export function DecorateAll(
+  decorator: <T>(
+    target: any,
+    propertyKey: string,
+    descriptor: TypedPropertyDescriptor<T>,
+  ) => TypedPropertyDescriptor<T> | void,
+) {
+  return (target: any) => {
+    const descriptors = Object.getOwnPropertyDescriptors(target.prototype);
+    for (const [propName, descriptor] of Object.entries(descriptors)) {
+      const isMethod = typeof descriptor.value == 'function' && propName !== 'constructor';
+      if (!isMethod) {
+        continue;
+      }
+      decorator({ ...target, constructor: { ...target.constructor, name: target.name } as any }, propName, descriptor);
+      Object.defineProperty(target.prototype, propName, descriptor);
+    }
+  };
+}
+
+export function ExecutionTimeHistogram({ description, unit = 'ms', valueType = ValueType.DOUBLE }: MetricOptions = {}) {
+  return (target: any, propertyKey: string | symbol, descriptor: PropertyDescriptor) => {
+    if (process.env.OTEL_SDK_DISABLED === 'true') {
+      return;
+    }
+
+    const method = descriptor.value;
+    const className = target.constructor.name as string;
+    const propertyName = String(propertyKey);
+    const metricName = `${_.snakeCase(className).replaceAll(/_(?=(repository)|(controller)|(provider)|(service)|(module))/g, '.')}.${_.snakeCase(propertyName)}.duration`;
+    const isAsync = method.constructor.name === 'AsyncFunction';
+
+    const metricDescription =
+      description ??
+      `The elapsed time in ${unit} for the ${_.startCase(className)} to ${_.startCase(propertyName).toLowerCase()}`;
+
+    let histogram: Histogram | undefined;
+
+    if (isAsync) {
+      descriptor.value = async function (...args: any[]) {
+        const start = performance.now();
+        const result = await method.apply(this, args);
+        const end = performance.now();
+        if (!histogram) {
+          histogram = metrics
+            .getMeter('immich')
+            .createHistogram(metricName, { description: metricDescription, unit, valueType });
+        }
+        histogram.record(end - start, {});
+        return result;
+      };
+    } else {
+      descriptor.value = function (...args: any[]) {
+        const start = performance.now();
+        const result = method.apply(this, args);
+        const end = performance.now();
+        if (!histogram) {
+          histogram = metrics
+            .getMeter('immich')
+            .createHistogram(metricName, { description: metricDescription, unit, valueType });
+        }
+        histogram.record(end - start, {});
+        return result;
+      };
+    }
```

Comment:
It bugs me how similar those two cases are but also I can't think of anything that could merge them

Reply:
Can you not check if the return type is an instance of a promise go from there?

Reply:
Have you seen line 168 and line 181?